nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSDocument.java

```diff
@@ -489,6 +501,70 @@ public void visitBlobs(Consumer<BlobAccessor> blobVisitor) throws PropertyExcept
         visitBlobs(docState.getState(), blobVisitor, docState::markDirty);
     }
 
+    @Override
+    public void makeRecord() {
+        DBSDocumentState docState = getStateOrTarget();
+        docState.put(KEY_IS_RECORD, TRUE);
+    }
+
+    @Override
+    public boolean isRecord() {
+        DBSDocumentState docState = getStateOrTarget();
+        return TRUE.equals(docState.get(KEY_IS_RECORD));
+    }
+
+    @Override
+    public void setRetainUntil(Calendar retainUntil) throws PropertyException {
+        DBSDocumentState docState = getStateOrTarget();
+        Calendar current = (Calendar) docState.get(KEY_RETAIN_UNTIL);
+        if (!allowNewRetention(current, retainUntil)) {
+            throw new PropertyException(
+                    "Cannot reduce retention time from: " + (current == null ? "null" : current.toInstant()) + " to: "
+                            + (retainUntil == null ? "null" : retainUntil.toInstant()));
+        }
+        docState.put(KEY_RETAIN_UNTIL, retainUntil);
```

Comment:
perhaps i am missed something, but if i understand correctly the concept of `record`, `retention` and `legal hold` are related. i means on javadoc of `makeRecord` we have `A record is a document with specific capabilities related to mandatory retention until a given date, and legal
     * holds....` but setting  the retention date / legal hold => should we or not make the document as a record  ?

Reply:
Yes we always have a choice to do these checks at the `CoreSession`/`DocumentModel` level or at the lower `Session`/`Document` level. I prefer to do the check asap at high level so that when we start calling events like BEFORE_SET_RETENTION we know things are ok. Then it's not worth duplicating the check at low level, given that the "record" aspects are just for convenience of being able to know when the blob has to be dispatched differently, but not really important from a data model level.